hippylib/algorithms/linalg.py

```diff
@@ -24,74 +25,84 @@ def amg_method(amg_type="ml_amg"):
     If not available use  petsc_amg.
     """
-    S = PETScKrylovSolver()
-    for pp in S.preconditioners():
+    for pp in dl.krylov_solver_preconditioners():
         if pp[0] == amg_type:
             return amg_type
         
     return 'petsc_amg'
 
-abspath = os.path.dirname( os.path.abspath(__file__) )
-source_directory = os.path.join(abspath,"cpp_linalg")
-header_file = open(os.path.join(source_directory,"linalg.h"), "r")
-code = header_file.read()
-header_file.close()
-cpp_sources = ["linalg.cpp"]  
-
-include_dirs = [".", source_directory]
-for ss in ['PROFILE_INSTALL_DIR', 'PETSC_DIR', 'SLEPC_DIR']:
-    if ss in os.environ.keys():
-        include_dirs.append(os.environ[ss]+'/include')
-        
-cpp_module = compile_extension_module(
-                code=code, source_directory=source_directory,
-                sources=cpp_sources, include_dirs=include_dirs)
-
 def MatMatMult(A,B):
     """
     Compute the matrix-matrix product :math:`AB`.
     """
-    s = cpp_module.cpp_linalg()
-    return s.MatMatMult(A,B)
+    Amat = dl.as_backend_type(A).mat()
+    Bmat = dl.as_backend_type(B).mat()
+    out = Amat.matMult(Bmat)
+    rmap, _ = Amat.getLGMap()
+    _, cmap = Bmat.getLGMap()
+    out.setLGMap(rmap, cmap)
+    return dl.Matrix(dl.PETScMatrix(out))
 
 def MatPtAP(A,P):
     """
     Compute the triple matrix product :math:`P^T A P`.
     """
-    s = cpp_module.cpp_linalg()
-    return s.MatPtAP(A,P)
+    Amat = dl.as_backend_type(A).mat()
+    Pmat = dl.as_backend_type(P).mat()
+    out = Amat.PtAP(Pmat, fill=1.0)
+    _, out_map = Pmat.getLGMap()
+    out.setLGMap(out_map, out_map)
+    return dl.Matrix(dl.PETScMatrix(out))
 
 def MatAtB(A,B):
     """
     Compute the matrix-matrix product :math:`A^T B`.
     """
-    s = cpp_module.cpp_linalg()
-    return s.MatAtB(A,B)
+    Amat = dl.as_backend_type(A).mat()
+    Bmat = dl.as_backend_type(B).mat()
+    out = Amat.transposeMatMult(Bmat)
+    _, rmap = Amat.getLGMap()
+    _, cmap = Bmat.getLGMap()
+    out.setLGMap(rmap, cmap)
+    return dl.Matrix(dl.PETScMatrix(out))
 
 def Transpose(A):
     """
     Compute the matrix transpose
     """
-    s = cpp_module.cpp_linalg()
-    return s.Transpose(A)
+    Amat = dl.as_backend_type(A).mat()
+    AT = PETSc.Mat()
+    Amat.transpose(AT)
+    rmap, cmap = Amat.getLGMap()
+    AT.setLGMap(cmap, rmap)
+    return dl.Matrix( dl.PETScMatrix(AT) )
 
 def SetToOwnedGid(v, gid, val):
-    s = cpp_module.cpp_linalg()
-    s.SetToOwnedGid(v, gid, val)
+    #    assert(v.owns_index(gid));
+    #    la_index index = static_cast<la_index>(gid);
+    #    v.set(&val, 1, &index);
+    # la_index_dtype
+    v[gid] = val
+
     
 def GetFromOwnedGid(v, gid):
-    s = cpp_module.cpp_linalg()
-    return s.GetFromOwnedGid(v, gid)
+    #    assert(v.owns_index(gid));
+    #    double val;
+    #    la_index index = static_cast<la_index>(gid);
+    #    v.get(&val, 1, &index);
+    #
+    #    return val;
```

Comment:
Some lines are commented out here, not sure if intentional or just forgot to delete

Reply:
You're right. I'll get rid of them.